bindings/rust/src/riscv.rs

```diff
@@ -41,41 +41,168 @@ pub enum RegisterRISCV {
+    INSTRET = 46,
+    HPMCOUNTER3 = 47,
+    HPMCOUNTER4 = 48,
+    HPMCOUNTER5 = 49,
+    HPMCOUNTER6 = 50,
+    HPMCOUNTER7 = 51,
+    HPMCOUNTER8 = 52,
+    HPMCOUNTER9 = 53,
+    HPMCOUNTER10 = 54,
+    HPMCOUNTER11 = 55,
+    HPMCOUNTER12 = 56,
+    HPMCOUNTER13 = 57,
+    HPMCOUNTER14 = 58,
+    HPMCOUNTER15 = 59,
+    HPMCOUNTER16 = 60,
+    HPMCOUNTER17 = 61,
+    HPMCOUNTER18 = 62,
+    HPMCOUNTER19 = 63,
+    HPMCOUNTER20 = 64,
+    HPMCOUNTER21 = 65,
+    HPMCOUNTER22 = 66,
+    HPMCOUNTER23 = 67,
+    HPMCOUNTER24 = 68,
+    HPMCOUNTER25 = 69,
+    HPMCOUNTER26 = 70,
+    HPMCOUNTER27 = 71,
+    HPMCOUNTER28 = 72,
+    HPMCOUNTER29 = 73,
+    HPMCOUNTER30 = 74,
+    HPMCOUNTER31 = 75,
+    CYCLEH = 76,
+    TIMEH = 77,
+    INSTRETH = 78,
+    HPMCOUNTER3H = 79,
+    HPMCOUNTER4H = 80,
+    HPMCOUNTER5H = 81,
+    HPMCOUNTER6H = 82,
+    HPMCOUNTER7H = 83,
+    HPMCOUNTER8H = 84,
+    HPMCOUNTER9H = 85,
+    HPMCOUNTER10H = 86,
+    HPMCOUNTER11H = 87,
+    HPMCOUNTER12H = 88,
+    HPMCOUNTER13H = 89,
+    HPMCOUNTER14H = 90,
+    HPMCOUNTER15H = 91,
+    HPMCOUNTER16H = 92,
+    HPMCOUNTER17H = 93,
+    HPMCOUNTER18H = 94,
+    HPMCOUNTER19H = 95,
+    HPMCOUNTER20H = 96,
+    HPMCOUNTER21H = 97,
+    HPMCOUNTER22H = 98,
+    HPMCOUNTER23H = 99,
+    HPMCOUNTER24H = 100,
+    HPMCOUNTER25H = 101,
+    HPMCOUNTER26H = 102,
+    HPMCOUNTER27H = 103,
+    HPMCOUNTER28H = 104,
+    HPMCOUNTER29H = 105,
+    HPMCOUNTER30H = 106,
+    HPMCOUNTER31H = 107,
+    MCYCLE = 108,
+    MINSTRET = 109,
+    MCYCLEH = 110,
+    MINSTRETH = 111,
+    MVENDORID = 112,
+    MARCHID = 113,
+    MIMPID = 114,
+    MHARTID = 115,
+    MSTATUS = 116,
+    MISA = 117,
+    MEDELEG = 118,
+    MIDELEG = 119,
+    MIE = 120,
+    MTVEC = 121,
+    MCOUNTEREN = 122,
+    MSTATUSH = 123,
+    MUCOUNTEREN = 124,
+    MSCOUNTEREN = 125,
+    MHCOUNTEREN = 126,
+    MSCRATCH = 127,
+    MEPC = 128,
+    MCAUSE = 129,
+    MTVAL = 130,
+    MIP = 131,
+    MBADADDR = 132,
+    SSTATUS = 133,
+    SEDELEG = 134,
+    SIDELEG = 135,
+    SIE = 136,
+    STVEC = 137,
+    SCOUNTEREN = 138,
+    SSCRATCH = 139,
+    SEPC = 140,
+    SCAUSE = 141,
+    STVAL = 142,
+    SIP = 143,
+    SBADADDR = 144,
+    SPTBR = 145,
+    SATP = 146,
+    HSTATUS = 147,
+    HEDELEG = 148,
+    HIDELEG = 149,
+    HIE = 150,
+    HCOUNTEREN = 151,
+    HTVAL = 152,
+    HIP = 153,
+    HTINST = 154,
+    HGATP = 155,
+    HTIMEDELTA = 156,
+    HTIMEDELTAH = 157,
+
     // Floating-point registers
-    F0 = 33,
-    F1 = 34,
-    F2 = 35,
-    F3 = 36,
-    F4 = 37,
-    F5 = 38,
-    F6 = 39,
-    F7 = 40,
-    F8 = 41,
-    F9 = 42,
-    F10 = 43,
-    F11 = 44,
-    F12 = 45,
-    F13 = 46,
-    F14 = 47,
-    F15 = 48,
-    F16 = 49,
-    F17 = 50,
-    F18 = 51,
-    F19 = 52,
-    F20 = 53,
-    F21 = 54,
-    F22 = 55,
-    F23 = 56,
-    F24 = 57,
-    F25 = 58,
-    F26 = 59,
-    F27 = 60,
-    F28 = 61,
-    F29 = 62,
-    F30 = 63,
-    F31 = 64,
-    PC = 65,
```

Comment:
Why not add the variables after `PC=65`?

Reply:
The conversion of `RegisterRISCV` in rust to `uc_riscv_reg` in C goes via an intermediate `int`. This means the conversion of `PC` should looks something like this:

```
RegisterRISCV::PC -> 190 -> UC_RISCV_REG_PC
```

However, this assumes that the mapping between `RegisterRISCV` and ints int rust (which is defined explicitly) matches that between `uc_riscv_reg` and ints in C (which is implicitly defined by the order of the enum). When the CSR registers were added to `uc_riscv_reg` the mapping in C became out of sync with the mapping in rust, so the conversion of `PC` from rust to C currently looks something like this:

```
RegisterRISCV::PC -> 65 -> UC_RISCV_REG_HPMCOUNTER21
```

As you see, the C side thinks we want to read HPMCOUNTER21, when we actually want to read the PC. This is obviously not good.

To fix this, we need to change the mappings in rust for all registers that have been shifted by the addition of the CSR registers in the C code.

As a side note: keeping the mappings between registers in C and rust synced by hand seems pretty error prone to me. Unfortunately I don't know a better alternative, because rust's FFI to my knowledge doesn't provide a mechanism to access definitions of enums in C.

Reply:
Sorry I realize it's my mistake [here](https://github.com/unicorn-engine/unicorn/commit/c6fdbb373506adc3c725c4e97f93c6953e53aca3). The variables are in the right place.